Dynamic `RegId` Support

DrChat opened the pull request:
### Description
As per a discussion in #53, ([comment](https://github.com/daniel5151/gdbstub/issues/53#issuecomment-849766528)), `gdbstub`'s `RegId` abstraction does not have a way to support dynamic register files due to the register size field returned from `RegId::from_raw_id`.

As suggested by @daniel5151, we can make the register size an optional hint, and enforce register sizing based on that.

This changeset also removes the array allocated in the single register access handler, and instead passes the target a closure that behaves in a similar manner to the one passed to `Registers::gdb_serialize`.

### API Stability

This will break SemVer compatibility, though for a good reason.

### Checklist

- Implementation
  - [x] `cargo build` compiles without `errors` or `warnings`
  - [x] `cargo clippy` runs without `errors` or `warnings`
  - [x] `cargo fmt` was run
  - [x] All tests pass
- Documentation
  - [ ] rustdoc + approprate inline code comments
  - [ ] Updated CHANGELOG.md
  - [ ] (if appropriate) Added feature to "Debugging Features" in README.md

### Validation

TODO...
P.S: How do I query a single register using a GDB client?

<!-- example output, from https://github.com/daniel5151/gdbstub/pull/54 -->

<details>
<summary>GDB output</summary>

```
TODO...
```

</details>

<details>
<summary>armv4t output</summary>

```
TODO...
```
</details>

DrChat commented:
```
PS C:\Dev\Repositories\gdbstub> cargo fmt --all -- --check
```

Not too sure what's up with that `cargo fmt` CI check... Mine is happy locally.

DrChat commented on gdbstub_arch/src/arm/reg/id.rs:
```diff
@@ -31,6 +33,6 @@ impl RegId for ArmCoreRegId {
             _ => return None,
         };
-        Some((reg, 4))
+        Some((reg, Some(NonZeroUsize::new(4).unwrap())))
```
Any recommendations for these `unwrap()` calls? I know they're not allowed in `gdbstub` code, but this will just get optimized out by the compiler.

daniel5151 replied:
I mean, since you're already in a function that returns `Option<_>`, you can just use `NonZeroUsize::new(x)?`.

At the end of the day, any reasonably optimizing compiler should generate equivalent code when using `?` or `unwrap` in these trivial cases (i.e: it'll elide the None branch entirely)

DrChat replied:
👍 using the `?` operator is a nicer shorthand :)
I wish there was a `new()` compile-time constructor that threw a compiler error if the constant evaluated to 0 - that'd be a bit nicer.

daniel5151 replied:
_sigh,_ preaching to the choir...

DrChat replied:
😂 time to submit some RFCs to Rust I guess

daniel5151 reviewed (changes requested):
Yeah, this change is looking pretty reasonable!

I've left some comments, but all in all, this seems to be on the right track.

I'll also pop open a `dev/0.6` branch which is based off the current `master`, and we'll switch the branch target for this PR over to it. EDIT: I just did that, so this PR will land in `dev/0.6`

daniel5151 commented:
With regards to `rustfmt`: `gdbstub` uses `rustfmt` nightly (for that sweet, sweet `wrap_comments` functionality), hence why the CI is getting mad. 

I should probably add that to the PR template...

* * *

As for how to read single registers from the GDB client... that's a bit tricky.

Here is the relevant code in GDB itself that fetches register values using the RSP: https://github.com/bminor/binutils-gdb/blob/master/gdb/remote.c#L8552-L8598

As you can see, it'll only fall back to using `fetch_register_using_p` (i.e: `read_register`) if the register isn't part of the standard `g` payload (i.e: `read_registers`). 

The motivating example that came up when this feature was first implemented was the RISC-V `Priv` register, which can only be queried via `read_register`. AFAIK, there isn't really an equivalent register like that in the ARMv4T arch, so I'm not sure if you'll be able to trigger this particular codepath via the `armv4t` example...

Instead, if there's a similar scenario that you can test in one of the architectures that your project uses, I would appreciate it if you could post the GDB output + protocol from your project. If not, that's _probably_ fine, since the diff is pretty small, and the implementation seems reasonable.

DrChat commented:
As always, thanks for the ⚡ fast review :)

I will work through your comments and make the appropriate changes.
Let me see if I can't figure out a way to test these changes with our target. For reference, I'm interested in the PowerPC OEA SPRs (things like the decrementer and such).
I won't be able to provide the test output unfortunately.

daniel5151 commented:
No problem! 
What better way to take a break from coding + reviewing PRs at work than by coding + reviewing PRs on your side projects 😋 

> I won't be able to provide the test output [...]

Unfortunate, but totally understandable. Since this is landing in the `dev/0.6` branch, I'm sure I'll circle back and give it a proper once-over at some point as well. That said, given how straightforward the changes are, I feel like things _should_ Just Work™️

DrChat commented:
Haha! Btw - heads up, I sent you an email at your gmail address with a couple of questions :)

DrChat commented on src/target/ext/base/single_register_access.rs:
```diff
@@ -35,7 +35,7 @@ pub trait SingleRegisterAccess<Id>: Target {
         tid: Id,
         reg_id: <Self::Arch as Arch>::RegId,
-        dst: &mut [u8],
+        output: &mut dyn FnMut(&[u8]),
```
Ah yeah - I was going to ask about this.
My initial thought was to just reuse the existing `Write` trait, since that's offered in `core`. Thoughts?

daniel5151 replied:
From the docs of the `core::fmt::Write` trait:

> A trait for writing or formatting into Unicode-accepting buffers or streams.

In this case, we are outputting raw binary data, so it wouldn't really be appropriate.

DrChat replied:
Shoot. I thought there was some sort of `io::Write` trait, but that might only be on `std` :(

daniel5151 commented:
Howdy! I'm back after a refreshing long weekend, and it's time to get back into the swing of things.

A cursory review of the PR indicates that there are still some comments that need to get resolved. Namely, there's still a matter of updating the documentation, replacing the current `unwrap` calls with `?` invocations, using a newtype instead of a bare callback, etc...

Since this is landing in `dev/0.6`, there's no rush to merge, but it certainly would be nice to close it out sooner rather than later :)

p.s: feel free to un-draft the PR

DrChat commented:
Appreciate it! Yeah, I still have some work to do with addressing those comments. Got distracted with other stuff so I just pushed off the changes I had :)

DrChat commented:
Going to need some thought on that latest commit - I appear to have hit an odd compiler error:
```
error: implementation of `FnOnce` is not general enough                                                                           
  --> src\gdbstub_impl\ext\single_register_access.rs:32:71
   |
32 |                 ops.read_register(id, reg_id, SendRegisterOutput::new(&mut send_output))
   |                                                                       ^^^^^^^^^^^^^^^^ implementation of `FnOnce` is not general enough
   |
   = note: closure with signature `fn(&'2 [u8])` must implement `FnOnce<(&'1 [u8],)>`, for any lifetime `'1`...
   = note: ...but it actually implements `FnOnce<(&'2 [u8],)>`, for some specific lifetime `'2`
```

I see the following issue on Rust that discusses this specific error: https://github.com/rust-lang/rust/issues/70263

daniel5151 commented:
Yeah, you're running into something related to https://doc.rust-lang.org/nomicon/hrtb.html

I'm a bit time strapped right now, but here's the fix:

```rust
/// An interface to send register data to the GDB remote debugger.
pub struct SendRegisterOutput<'a> {
    inner: &'a mut dyn for<'b> FnMut(&'b [u8]),
}

impl<'a> SendRegisterOutput<'a> {
    pub(crate) fn new(inner: &'a mut dyn for<'b> FnMut(&'b [u8])) -> Self {
        Self { inner }
    }

    /// Write out raw register bytes to the GDB debugger.
    pub fn write(&mut self, data: &[u8]) {
        (self.inner)(data)
    }
}

// and then later, in the implementation, DONT use separate named closure variable

let mut err = Ok(());

// TODO: Limit the number of bytes transferred on the wire to the register size
// (if specified).
ops.read_register(
    id,
    reg_id, 
    SendRegisterOutput::new(&mut |buf| match res.write_hex_buf(buf) {
        Ok(_) => {}
        Err(e) => err = Err(e),
    }),
)
.handle_error()?;

err?;
```

DrChat commented:
¯\_(ツ)_/¯ Well, the compiler was probably right but I don't know why.

daniel5151 commented on src/target/ext/base/single_register_access.rs:
```diff
@@ -35,7 +35,7 @@ pub trait SingleRegisterAccess<Id>: Target {
         &mut self,
```
meta: you'll also need to update the `read_register` docs to rephrase the following into something more correct:

```
/// Implementations should write the value of the register using target's
/// native byte order in the buffer `dst`
```

daniel5151 commented on src/arch.rs:
```diff
@@ -28,15 +28,20 @@ use crate::internal::{BeBytes, LeBytes};
+    /// Map raw GDB register number to a corresponding `RegId` and optional
+    /// register size.
+    ///
+    /// If the register size is specified here, no more than that amount of
+    /// bytes will be permitted to be transferred on the wire for that
+    /// register.
```
```suggestion
    /// If the register size is specified here, gdbstub will include a runtime check that 
    /// ensures target implementations do not send back more bytes than the register allows
```

daniel5151 reviewed (changes requested):
You've also still got to swap those `.unwrap()` calls for `?` invocations.

daniel5151 commented:
If you'd like to add a SPR to the armv4t example, that would be great! That said, I'm not sure how easy it'll be, and it might require tweaking the underling `Arch` implementation, which can be a bit of a hassle...

If not, then let me know once you've tested it and confirmed that things are working as expected.

DrChat commented:
Ack. It appears that GDB does not expose `SPSR` (for some reason).
I will have to do some testing on my own, I'll report back later.

daniel5151 commented:
> P.S: Can we put down some links to GDB source above the `from_raw_id` implementations?

That's a good idea. It shouldn't be _too_ hard, since IIRC they should just be the same links as those linked to by the associated `Arch`. Up to you if you'd like to do it as part of this PR, or if we should just punt it to a follow up task that someone can tackle at some point.

DrChat commented:
Perfect - I'll do that as part of this PR!

Also, it looks like GDB prefers the `P` packet when writing registers, which you can trigger with:
```
set $r31 = 0x1
```

daniel5151 commented:
> Also, it looks like GDB prefers the `P` packet when writing registers

Yep, I know. Too bad we aren't actually modifying the `P` packet in this PR ¯\\\_(ツ)_/¯

daniel5151 commented on src/arch.rs:
```diff
@@ -28,15 +28,20 @@ use crate::internal::{BeBytes, LeBytes};
+    ///
+    /// If the register size is specified here, gdbstub will include a runtime
+    /// check that ensures target implementations do not send back more
+    /// bytes than the register allows
```
nit: missing period :P

DrChat commented:
True - just a little note for myself so I can use it to test `from_raw_id` :)

daniel5151 commented on src/gdbstub_impl/ext/single_register_access.rs:
```diff
@@ -13,17 +13,40 @@ impl<T: Target, C: Connection> GdbStubImpl<T, C> {
+                            // If the register has a known size and read_register attempts
+                            // to send more bytes than are present in the register,
+                            // error out and stop sending data.
+                            n += buf.len();
```
move this line + the set error line into the `if let Some` branch. If `from_raw_id` is inlined + always returns `Some(usize, None)`, then the entire contents of the callback can be DCEd

DrChat replied:
👀 which error line? I can't see how to move the `err = res.write_hex_buf ...` line.

daniel5151 replied:
ah, uhh, nevermind 😅
coffee hasn't fully kicked in yet

daniel5151 commented:
Any updates on this PR? Did you have a chance to validate that things are working as expected?

daniel5151 commented on src/target/ext/base/single_register_access.rs:
```diff
@@ -65,3 +65,19 @@ pub trait SingleRegisterAccess<Id>: Target {
 pub type SingleRegisterAccessOps<'a, Id, T> =
     &'a mut dyn SingleRegisterAccess<Id, Arch = <T as Target>::Arch, Error = <T as Target>::Error>;
+
+/// An interface to send register data to the GDB remote debugger.
```
```suggestion
/// An interface to send register data back to the GDB client.
```

daniel5151 commented on src/target/ext/base/single_register_access.rs:
```diff
@@ -65,3 +65,19 @@ pub trait SingleRegisterAccess<Id>: Target {
+        Self { inner }
+    }
+
+    /// Write out raw register bytes to the GDB debugger.
```
```suggestion
    /// Write raw register bytes to GDB. Bytes must be sent in the target's native endian order.
```

GDB stands for GNU DeBugger, so "GDB debugger" is redundant (like ATM Machine)

daniel5151 reviewed (changes requested):
a couple of minor doc tweaks

DrChat commented:
Hey - no updates as of now. I got hung up on the issue that gdb does not actually expose SPRs for common PPC architectures (even OEA SPRs) and went to different tasking.

I can apply your suggested changes on Monday and then I probably won't be working on this for a few weeks.

daniel5151 commented:
Gotcha. All good!

FYI, it seems that 0.6 will be coming out a bit sooner than expected (thanks to a recently reported bug in the breakpoint/watchpoints API + the in-progress state-machine refactoring I've been working on), so there's a non-zero chance I'll just pull the trigger and merge this PR as-is at some point in the coming weeks. I think these changes are worthwhile, and I'd rather have them land as part of 0.6, rather than some future 0.7 release.

daniel5151 commented:
Oh, and P.S: you _should_ be able to manually define custom SPRs by returning a custom target.xml. Just because all the architectures currently implemented in `gdbstub_arch` stick to using the built-in target.xml features that ship with GDB, doesn't mean that you aren't able to define your own target-specific registers (at least in theory).

See https://sourceware.org/gdb/current/onlinedocs/gdb/Target-Description-Format.html#Target-Description-Format for more details (with special attention to `G.2.5 Features` and `G.2.6 Types`

DrChat commented:
Oh - huh, I never thought of that!
I'll do more research into that. Thanks for letting me know!

daniel5151 commented:
Hey, heads up, I decided to see what exactly defining a custom `target.xml` file would entail, and before I knew it, I ended up adding a custom `target.xml` to the `armv4t` example. See commit 5562eec

If you run the armv4t example now, you'll see two new non-standard registers show up: `custom` and `time`. The latter of which (`time`) is set up such that it doesn't get sent back to the GDB stub as part of the regular `g` packet response, which forces the stub to read it via the `p` packet 😄 

Hopefully this new example helps you out with defining custom target.xml files. Also, you'll need to rebase this PR on-top of these changes, and make sure that the 'p' packet still works as intended.

Cheers

DrChat commented:
Ah - perfect! Thanks for looking into that!
Sorry I haven't been putting much time into this - I've been working unrelated stuff lately.

daniel5151 commented:
Yeah, no worries. Playing around with custom target.xml files is something I've been meaning to do for a while now :)

bet4it commented:
Hope this PR will be merged in `0.6`

daniel5151 commented:
Yep, I'll formally add it to the roadmap, and I'll look into merging it into dev/0.6 sometime soon-ish.

I've been a bit busy lately, and haven't had too much time to work on gdbstub. Hopefully I'll have some time over the July 4th long weekend to polish up some changes I've been working on + integrate these changes. Not sure about a hard release date for 0.6 yet, but I'd wager it's still a few weeks away.

bet4it commented:
Any updates?

daniel5151 commented:
As you might have guessed, my plan to "polish up my local changes + integrate these changes" hadn't panned out as I hoped 😅 

I've been quite busy with work / life recently, so while I've been able to carve out some time in the morning to review PRs, I haven't had the chance to get back into "hands-on programming" mode with `gdbstub`. 

With the vFile stuff coming in though, `dev/0.6` will have _yet more_ genuinely useful changes, so i'll really need to block out some time where I can hammer things out...
